macos/resource_compression/dcmp_2.ksy

```diff
@@ -0,0 +1,289 @@
+meta:
+  id: dcmp_2
+  title: Compressed Macintosh resource data, Apple `'dcmp' (2)` format
+  application: Mac OS
+  license: MIT
+  ks-version: "0.8"
+  imports:
+    - /common/bytes_with_io
+  endian: be
+doc: |
+  Compressed resource data in `'dcmp' (2)` format,
+  as stored in compressed resources with header type `9` and decompressor ID `2`.
+
+  The `'dcmp' (2)` decompressor resource is included in the System file of System 7.0 and later.
+  This compression format is used for a few compressed resources in System 7.0's files
+  (such as the System file).
+  This decompressor is also included with and used by some other Apple applications,
+  such as ResEdit.
+  (Note: ResEdit includes the `'dcmp' (2)` resource,
+  but none of its resources actually use this decompressor.)
+
+  This compression format is based on simple dictionary coding,
+  where each byte in the compressed data expands to two bytes,
+  based on a lookup table
+  (either included in the compressed data or provided by the decompressor).
+  An alternative "tagged" compression format is also supported,
+  which allows using two-byte literals in addition to single-byte table references,
+  at the cost of requiring an extra "tag" byte every 16 output bytes,
+  to differentiate literals and table references.
+doc-ref: 'https://github.com/dgelessus/python-rsrcfork/tree/master/rsrcfork/compress/dcmp2.py'
+params:
+  - id: len_decompressed
+    type: u4
+    doc: |
+      The length of the decompressed data in bytes,
+      from the compressed resource header.
+  - id: header_parameters_with_io
+    type: bytes_with_io
+    doc: |
+      The unparsed decompressor-specific parameters,
+      from the compressed resource header.
+seq:
+  - id: custom_lookup_table
+    size: 2
+    repeat: expr
+    repeat-expr: header_parameters.num_custom_lookup_table_entries
+    if: header_parameters.flags.has_custom_lookup_table
+    doc: |
+      The custom lookup table to be used instead of the default lookup table.
+  - id: data
+    type:
+      switch-on: header_parameters.flags.tagged
+      cases:
+        false: untagged_data
+        true: tagged_data
```

Comment:
I'd personally prefer this variant:

```suggestion
      cases:
        true: tagged_data
        _: untagged_data
```

..., because then the "default byte array" case isn't generated at all and the byte array type also doesn't pollute the resulting combined type in static typed languages (combining two different user types with a byte array type results in generic `Object` in Java, whereas combining only two user types yields `KaitaiStruct`). If any attribute has size and doesn't have the else case denoted by underscore `_`, the compiler logically assumes that the cases do not fill the full range of possible values the expression in `switch-on` can have and thus generates the fallback byte array case.

But it's rather such a side note, if you think that having verbatim values `true` and `false` helps readability, I accept it. The type casting in statically-typed languages will still be needed in most cases because the result type would be `KaitaiStruct` (unless the programmer wants to only access properties available in `KaitaiStruct` directly like `_io` or `_parent`), but it can reduce confusion of why this attribute has type `Object` when it should apparently be `KaitaiStruct`.

Reply:
Makes sense. It's really annoying that the compiler behaves this way - IMO the default behavior of `switch-on` should be to generate an error if a value is encountered that isn't listed under cases (and there is no explicit `_` branch). Of course this can't be changed easily because of backwards compatibility. Also, there's no explicit name for the byte array type, so the implicit default case is actually necessary if you want to have a byte array case in a `switch-on`...

kaitai-io/kaitai_struct#208 would be useful to have here.